File: Difficulty: Medium/Rotten Oranges/rotten-oranges.cpp
class Solution {
    private:
    
    bool isSafe(vector<vector<int>>& mat, int r, int c){
        
        return (r >=0 && r < mat.size() && c >= 0 && c < mat[0].size() && mat[r][c]==1);
    }
    
  public:
    int orangesRot(vector<vector<int>>& mat) {
        // code here
        queue<pair<int,int>> q;
        int rowSize = mat.size(), colSize = mat[0].size();
        int freshOranges = 0;
        //level order traversal 
        // i need all the rotten oranges at the same time inside queue
        for(int i=0;i<rowSize; i++){
            for(int j=0; j< colSize;j++){
                //rotten orange
                if(mat[i][j]==2){
                    q.push({i,j});
                }else if(mat[i][j]==1) freshOranges++;
            }
        }
        
        if(freshOranges ==0) return 0;
        
        int totalTime =0;
        
        //bfs on all the rotten eggs at the same time in 4 directions 
        vector<int> dRow = {-1,1,0,0};
        vector<int> dCol = {0,0,-1,1};
        
        
        while(!q.empty()){
            
            int Size = q.size() ; 
            bool rotted = false;
            
            for(int i=0;i< Size; i++){
                
                auto [r,c] = q.front();
                q.pop();
                
                for(int j=0;j<4;j++){
                    //check for all four directions 
                    int nRow = r + dRow[j];
                    int nCol = c + dCol[j];
                    
                    if(isSafe(mat,nRow,nCol)){
                        mat[nRow][nCol]=2;
                        freshOranges--;
                        q.push({nRow,nCol});
                        rotted = true;
                    }
                }
            }
            if(rotted) totalTime++;
        }
        
        return freshOranges  > 0 ? -1 : totalTime;
    }
};

File: Difficulty: Medium/Rotten Oranges/rotten-oranges.py

from queue import Queue

class Solution:
	def orangesRot(self, mat):
		# code here
		
		time = -1
		dir = [(-1,0),(1,0),(0,-1),(0,1)]
		n, m = len(mat), len(mat[0])
		q = Queue()
		fresh = 0
		
		for i,r in enumerate(mat):
		    for j,val in enumerate(r):
		        if val == 2:
		            q.put((i,j))
		        elif val == 1:
		            fresh +=1 
		            
		if fresh == 0:
		    return 0
		     
		while not q.empty():
		    size = q.qsize()
		    while size > 0:
		        curr_i, curr_j = q.get()
		        for direction in dir:
		            
		            next_i = direction[0]
		            next_j = direction[1]
		             
		            new_i = curr_i + next_i
		            new_j = curr_j + next_j
		             
		            # check if the new index is inbound and is a 1
		            if new_i < n and new_i >=0 and new_j >=0 and new_j < m and mat[new_i][new_j] == 1:
		                mat[new_i][new_j] = 2
		                fresh -=1
		                q.put((new_i, new_j))
		               
                size -=1
		    time += 1
		
		  
	    return time if fresh == 0 else -1
		     
		 
		        